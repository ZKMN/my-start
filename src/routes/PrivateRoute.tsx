import React from 'react';
import { connect } from 'react-redux';
import {
  Route,
  Redirect,
  RouteProps,
  RouteComponentProps,
} from 'react-router-dom';

import { RootState } from 'index';
import Links from 'links';

interface IPrivateRoute extends RouteProps {
  isLoggedIn: boolean;
  user: Record<string, unknown> | null;
  component: React.FC<RouteComponentProps>
}

export const PrivateRoute = ({
  component: Component,
  isLoggedIn,
  user,
  path,
  ...rest
}: IPrivateRoute) => (
  <Route
    {...rest}
    path={path}
    render={({
      location,
      match,
      history,
    }) => {
      if (!isLoggedIn && !user) {
        // not loggedIn - redirect to login page
        return (
          <Redirect
            to={{
              pathname: Links.Login,
              state: { redirectFrom: location.pathname },
            }}
          />
        );
      }

      return (
        <Component
          key={location.key} // leave key for update component when you click on same link in navigaton
          location={location}
          match={match}
          history={history}
        />
      );
    }}
  />
);

export default connect((state: RootState) => ({
  isLoggedIn: state.login.isLoggedIn,
  user: state.login.user,
}))(PrivateRoute);
