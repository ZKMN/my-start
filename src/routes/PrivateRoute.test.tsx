import React from 'react';
import { shallow } from 'enzyme';
import { Route } from 'react-router-dom';
import { PrivateRoute } from './PrivateRoute';

let wrapper: React.ReactElement;

const props = {
  component: () => 'some comp',
  path: '/',
};

describe('PrivateRoute', () => {
  beforeEach(() => {
    wrapper = shallow(<PrivateRoute {...props} />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('matches snapshot if isLoggedIn && commonDataLoaded && user && !user.hasFullAccess', () => {
    wrapper.setProps({
      user: { hasFullAccess: false },
      isLoggedIn: true,
      commonDataLoaded: true,
    });

    const result = wrapper
      .find(Route)
      .at(0)
      .props()
      .render({
        location: {},
        match: {},
        history: {},
      });

    expect(result).toMatchSnapshot();
  });

  it('matches snapshot if isLoggedIn && user && !commonDataLoaded', () => {
    wrapper.setProps({
      isLoggedIn: true,
      commonDataLoaded: false,
      user: { hasFullAccess: true },
    });

    const result = wrapper
      .find(Route)
      .at(0)
      .props()
      .render({
        location: {},
        match: {},
        history: {},
      });

    expect(result).toMatchSnapshot();
  });

  it('matches snapshot if !isLoggedIn && !user', () => {
    wrapper.setProps({
      isLoggedIn: false,
      user: null,
    });

    const result = wrapper
      .find(Route)
      .at(0)
      .props()
      .render({
        location: {},
        match: {},
        history: {},
      });

    expect(result).toMatchSnapshot();
  });
});
