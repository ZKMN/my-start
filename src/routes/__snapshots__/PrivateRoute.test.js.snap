// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute matches snapshot if !isLoggedIn && !user 1`] = `
<Redirect
  to={
    Object {
      "pathname": "/login",
      "state": Object {
        "redirectFrom": undefined,
      },
    }
  }
/>
`;

exports[`PrivateRoute matches snapshot if isLoggedIn && commonDataLoaded && user && !user.hasFullAccess 1`] = `
<component
  history={Object {}}
  location={Object {}}
  match={Object {}}
/>
`;

exports[`PrivateRoute matches snapshot if isLoggedIn && user && !commonDataLoaded 1`] = `
<Redirect
  to={
    Object {
      "pathname": "/loading",
      "state": Object {
        "redirectFrom": undefined,
        "redirectFromSearch": undefined,
      },
    }
  }
/>
`;
